src/column-layout/with-css.tsx

```diff
@@ -0,0 +1,86 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import React from 'react';
+import clsx from 'clsx';
+import flattenChildren from 'react-keyed-flatten-children';
+import { useContainerQuery } from '../internal/hooks/container-queries';
+import { getCommonClasses } from './util';
+import { InternalColumnLayoutProps } from './interfaces';
+import customCssProps from '../internal/generated/custom-css-properties';
+import styles from './styles.css.js';
+
+const isOdd = (value: number): boolean => value % 2 !== 0;
+
+export function calculcateCssColumnCount(
+  columns: number,
+  minColumnWidth: number,
+  containerWidth: number | null
+): number {
+  if (!containerWidth) {
+    return columns;
+  }
+
+  // First, calculate how many columns we can have based on the current container width and minColumnWidth.
+  const targetColumnCount = Math.min(columns, Math.floor(containerWidth / minColumnWidth));
+
+  // When we start wrapping into fewer columns than desired, we want to keep the number of columns even.
+  return Math.max(
+    1,
+    targetColumnCount < columns && isOdd(targetColumnCount) ? targetColumnCount - 1 : targetColumnCount
+  );
+}
+
+interface ColumnLayoutWithCSSProps
+  extends Pick<InternalColumnLayoutProps, 'minColumnWidth' | 'columns' | 'variant' | 'borders' | 'disableGutters'> {
+  children: React.ReactNode;
+}
+
+export default function ColumnLayoutWithCSS({
+  columns = 1,
+  minColumnWidth = 0,
+  variant,
+  borders,
+  disableGutters,
+  children,
+}: ColumnLayoutWithCSSProps) {
+  const [containerWidth, containerRef] = useContainerQuery(rect => rect.width);
+
+  const columnCount = calculcateCssColumnCount(columns, minColumnWidth, containerWidth);
+
+  // Flattening the children allows us to "see through" React Fragments and nested arrays.
+  const flattenedChildren = flattenChildren(children);
+  const childrenCount = flattenedChildren.length;
+  const rowCount = Math.ceil(childrenCount / columnCount);
+
+  return (
+    <div
+      ref={containerRef}
+      className={clsx(
+        ...getCommonClasses({ borders, disableGutters, variant }),
+        styles['css-grid'],
+        isOdd(columnCount) && styles['odd-columns']
+      )}
+      style={{ [customCssProps.columnLayoutColumnCount]: columnCount }}
+    >
+      {flattenedChildren.map((child, i) => {
+        // If this react child is a primitive value, the key will be undefined
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const key = (child as any).key;
+        const rowIndex = Math.floor(i / columnCount);
+
+        return (
+          <div
+            key={key}
+            className={clsx(styles['restore-pointer-events'], {
```

Comment:
Do we still need the restore-pointer-events trick in this implementation?

Reply:
Why not grid-gap? Unification with the other layout? Same idea as https://github.com/cloudscape-design/components/pull/1123#discussion_r1217886392 maybe it is worth separating two instances properly.

Reply:
Yep, I've thought about this a lot and decided to go with this more unified and conservative approach. Thank you for the additional data point though. I will think this over again. I would really like to remove this.